File: src/renderer/src/smartSilence/presets.ts
// Smart Silence Cleaner — the five editing modes (presets).
//
// Faithful port of smart_silence_cleaner/models/presets.py. The EXACT numbers
// come straight from the product spec so the modes behave predictably.

import type { Preset } from './types'
import { copySettings } from './settings'

export const CONSERVATIVE: Preset = {
  id: 'conservative',
  icon: '🐢',
  title: 'Keep More Natural Pauses',
  description:
    'Perfect for storytelling, interviews, podcasts, emotional videos and cinematic pacing. Very light trimming.',
  editable: false,
  settings: {
    minGapMs: 450,
    targetPauseMs: 280,
    sentencePauseMs: 450,
    commaPauseMs: 300,
    mergeGapMs: 140,
    padBeforeMs: 90,
    padAfterMs: 110,
    keepSentenceFlow: true,
    removeFillers: false,
    smartSilence: true,
  },
}

export const BALANCED: Preset = {
  id: 'balanced',
  icon: '⚖️',
  title: 'Recommended',
  description: 'Removes awkward dead air while keeping speech natural. This is the default.',
  editable: false,
  settings: {
    minGapMs: 300,
    targetPauseMs: 180,
    sentencePauseMs: 350,
    commaPauseMs: 250,
    mergeGapMs: 100,
    padBeforeMs: 60,
    padAfterMs: 80,
    keepSentenceFlow: true,
    removeFillers: false,
    smartSilence: true,
  },
}

export const AGGRESSIVE: Preset = {
  id: 'aggressive',
  icon: '⚡',
  title: 'Fast Paced',
  description: 'Creates energetic YouTube Shorts, Reels and TikTok-style pacing.',
  editable: false,
  settings: {
    minGapMs: 180,
    targetPauseMs: 80,
    sentencePauseMs: 220,
    commaPauseMs: 150,
    mergeGapMs: 80,
    padBeforeMs: 40,
    padAfterMs: 50,
    keepSentenceFlow: true,
    removeFillers: false,
    smartSilence: true,
  },
}

// Strip *every* gap over 80 ms down to nothing — a true gapless cut. Smart
// Silence is off (uniform, no punctuation rules), pads are zero.
export const AGGRESSIVE_GAPLESS: Preset = {
  id: 'aggressive_gapless',
  icon: '✂️',
  title: 'Aggressive Gapless',
  description: 'Removes every silence gap longer than 80 ms. Tightest possible — no breathing room.',
  editable: false,
  settings: {
    minGapMs: 80,
    targetPauseMs: 0,
    sentencePauseMs: 0,
    commaPauseMs: 0,
    mergeGapMs: 60,
    padBeforeMs: 0,
    padAfterMs: 0,
    keepSentenceFlow: false,
    removeFillers: false,
    smartSilence: false,
  },
}

// Custom starts as a copy of Balanced and is the one mode the user may tune.
export const CUSTOM: Preset = {
  id: 'custom',
  icon: '🎚️',
  title: 'Custom',
  description: 'Adjust everything manually with the sliders below, or open Advanced for exact values.',
  editable: true,
  settings: copySettings(BALANCED.settings),
}

/** Display order for the mode cards. */
export const PRESET_ORDER: readonly Preset[] = [
  CONSERVATIVE,
  BALANCED,
  AGGRESSIVE,
  AGGRESSIVE_GAPLESS,
  CUSTOM,
]

/** The mode selected on first open. */
export const DEFAULT_PRESET_ID = 'balanced'

const BY_ID: ReadonlyMap<string, Preset> = new Map(PRESET_ORDER.map((p) => [p.id, p]))

/** Every preset, in display order. */
export function allPresets(): readonly Preset[] {
  return PRESET_ORDER
}

/** Look up a preset by id (falls back to the default). */
export function getPreset(id: string): Preset {
  return BY_ID.get(id) ?? BY_ID.get(DEFAULT_PRESET_ID)!
}

/** The Balanced/Recommended default. */
export function defaultPreset(): Preset {
  return getPreset(DEFAULT_PRESET_ID)
}

File: smart_silence_cleaner/models/presets.py
"""The editing modes (presets) and the manager that hands them out.

Each preset bundles creator-facing copy (icon / title / description) with the
raw :class:`SilenceSettings` it stands for. The exact numbers come straight from
the product spec so the modes behave predictably.
"""

from __future__ import annotations

from dataclasses import dataclass

from .settings import SilenceSettings


@dataclass(frozen=True, slots=True)
class Preset:
    """A named editing mode shown as a large card in the UI."""

    id: str
    icon: str
    title: str
    description: str
    settings: SilenceSettings
    #: Custom is the only mode whose sliders/advanced params are user-editable.
    editable: bool = False


# --------------------------------------------------------------------------- #
# The five modes                                                              #
# --------------------------------------------------------------------------- #

CONSERVATIVE = Preset(
    id="conservative",
    icon="🐢",
    title="Keep More Natural Pauses",
    description=(
        "Perfect for storytelling, interviews, podcasts, emotional videos and "
        "cinematic pacing. Very light trimming."
    ),
    settings=SilenceSettings(
        min_gap_ms=450,
        target_pause_ms=280,
        sentence_pause_ms=450,
        comma_pause_ms=300,
        merge_gap_ms=140,
        pad_before_ms=90,
        pad_after_ms=110,
        keep_sentence_flow=True,
        remove_fillers=False,
        smart_silence=True,
    ),
)

BALANCED = Preset(
    id="balanced",
    icon="⚖️",
    title="Recommended",
    description="Removes awkward dead air while keeping speech natural. This is the default.",
    settings=SilenceSettings(
        min_gap_ms=300,
        target_pause_ms=180,
        sentence_pause_ms=350,
        comma_pause_ms=250,
        merge_gap_ms=100,
        pad_before_ms=60,
        pad_after_ms=80,
        keep_sentence_flow=True,
        remove_fillers=False,
        smart_silence=True,
    ),
)

AGGRESSIVE = Preset(
    id="aggressive",
    icon="⚡",
    title="Fast Paced",
    description="Creates energetic YouTube Shorts, Reels and TikTok-style pacing.",
    settings=SilenceSettings(
        min_gap_ms=180,
        target_pause_ms=80,
        sentence_pause_ms=220,
        comma_pause_ms=150,
        merge_gap_ms=80,
        pad_before_ms=40,
        pad_after_ms=50,
        keep_sentence_flow=True,
        remove_fillers=False,
        smart_silence=True,
    ),
)

# Extra mode requested on top of the spec: strip *every* gap over 80 ms down to
# nothing — a true gapless cut. Smart Silence is off (uniform, no punctuation
# rules) and the pads are zero so words butt right together.
AGGRESSIVE_GAPLESS = Preset(
    id="aggressive_gapless",
    icon="✂️",
    title="Aggressive Gapless",
    description="Removes every silence gap longer than 80 ms. Tightest possible — no breathing room.",
    settings=SilenceSettings(
        min_gap_ms=80,
        target_pause_ms=0,
        sentence_pause_ms=0,
        comma_pause_ms=0,
        merge_gap_ms=60,
        pad_before_ms=0,
        pad_after_ms=0,
        keep_sentence_flow=False,
        remove_fillers=False,
        smart_silence=False,
    ),
)

# Custom starts as a copy of Balanced and is the one mode the user may tune.
CUSTOM = Preset(
    id="custom",
    icon="🎚️",
    title="Custom",
    description="Adjust everything manually with the sliders below, or open Advanced for exact values.",
    settings=BALANCED.settings.copy(),
    editable=True,
)


class PresetManager:
    """Ordered access to the editing modes."""

    #: Display order for the mode cards.
    ORDER: tuple[Preset, ...] = (CONSERVATIVE, BALANCED, AGGRESSIVE, AGGRESSIVE_GAPLESS, CUSTOM)

    #: The mode selected on first open.
    DEFAULT_ID = "balanced"

    _BY_ID = {p.id: p for p in ORDER}

    @classmethod
    def all(cls) -> tuple[Preset, ...]:
        """Every preset, in display order."""
        return cls.ORDER

    @classmethod
    def get(cls, preset_id: str) -> Preset:
        """Look up a preset by id (falls back to the default)."""
        return cls._BY_ID.get(preset_id, cls._BY_ID[cls.DEFAULT_ID])

    @classmethod
    def default(cls) -> Preset:
        """The Balanced/Recommended default."""
        return cls.get(cls.DEFAULT_ID)
